### PR TITLE
Added SVM Classifier

### LINKED ISSUE #99: Add SVM classifier for existing metrics

### PR DESCRIPTION
**Pull request by John42506176Linux**
SVM classifier has been fully added with probability for f1 score , precision and recall metrics.
SVMtestoutput
SVMoutput

Closes #99

### CODE REVIEW

File: python-package/brainflow/ml_model.py [hunk 1]
```diff
@@ -27,7 +27,8 @@ class BrainFlowClassifiers (enum.Enum):
     """Enum to store all supported classifiers"""
 
     REGRESSION = 0 #:
-    KNN = 1 #:
+    KNN = 1  #:
```
Comment (Andrey1994) [inline python-package/brainflow/ml_model.py:30]: extra space here

File: src/ml/concentration_svm_classifier.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+#include <algorithm>
+#include <cmath>
+#include <stdlib.h>
+#include <iostream>
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:4]: dont need iostream

File: src/ml/concentration_svm_classifier.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#include "brainflow_constants.h"
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:5]: lets add a new line before local includes

File: src/ml/concentration_svm_classifier.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#include "concentration_svm_classifier.h"
+#include "get_dll_dir.h"
+
+int ConcentrationSVMClassifier::prepare ()
+{
+    char path[1024];
+    bool res = get_dll_path (path);
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:13]: this string looks like clang-format doesnt work

File: src/ml/concentration_svm_classifier.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    std::strcpy (fullPath,path);
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:14]: as well as here and below, I belive clang-format adds a space after comma

File: src/ml/concentration_svm_classifier.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    std::strcat (fullPath ,"brainflow_svm.model");
+    model = svm_load_model (fullPath);
+    return (int)BrainFlowExitCodes::STATUS_OK;
+}
+
+int ConcentrationSVMClassifier::predict (double *data, int data_len, double *output)
+{
+    if ((data_len < 5) || (data == NULL) || (output == NULL))
+    {
+        safe_logger (spdlog::level::err, "Classifier has already been prepared.");
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:24]: log message is wrong

File: src/ml/concentration_svm_classifier.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
+    std::strcpy (fullPath,path);
+    std::strcat (fullPath ,"brainflow_svm.model");
+    model = svm_load_model (fullPath);
+    return (int)BrainFlowExitCodes::STATUS_OK;
+}
+
+int ConcentrationSVMClassifier::predict (double *data, int data_len, double *output)
+{
+    if ((data_len < 5) || (data == NULL) || (output == NULL))
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:22]: does it work if data_len is 5 instead 10?

File: src/ml/concentration_svm_classifier.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        safe_logger (spdlog::level::err, "Classifier has already been prepared.");
+        return (int)BrainFlowExitCodes::INVALID_BUFFER_SIZE_ERROR;
+    }
+    struct svm_node *x;
+    x =  (struct svm_node *) malloc(data_len*sizeof(struct svm_node));
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:28]: looks like memory leak, I dont see free call in the end. You can try to run it with valgrind to test

Comment (Andrey1994) [reply src/ml/concentration_svm_classifier.cpp:28]: and clang-format is definitely off on your PC

File: src/ml/concentration_svm_classifier.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    svm_predict_probability (model, x,prob_estimates);
+    *output = prob_estimates[0];
+    delete [] x;
+    delete [] prob_estimates;
+    return (int)BrainFlowExitCodes::STATUS_OK;
+}
+
+int ConcentrationSVMClassifier::release ()
+{
+    delete model;
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:46]: add a check that its not NULL or that classifier is prepared

File: src/ml/concentration_svm_classifier.cpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+#include "get_dll_dir.h"
+
+int ConcentrationSVMClassifier::prepare ()
+{
+    char path[1024];
+    bool res = get_dll_path (path);
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
+    std::strcpy (fullPath,path);
+    std::strcat (fullPath ,"brainflow_svm.model");
+    model = svm_load_model (fullPath);
```
Comment (Andrey1994) [inline src/ml/concentration_svm_classifier.cpp:16]: we use _ in naming instead camel case

File: tests/cpp/ml_demo/src/eeg_metrics.cpp [hunk 11]
```diff
@@ -11,6 +11,7 @@
 #include "board_shim.h"
 #include "data_filter.h"
 #include "ml_model.h"
+#include  <vector>
```
Comment (Andrey1994) [inline tests/cpp/ml_demo/src/eeg_metrics.cpp:14]: can we move SVM to another test\file? This example is used in docs on public page as code sample. I want to keep it short and simple for demo purposes

File: tests/python/eeg_metrics.py [hunk 12]
```diff
@@ -38,9 +39,14 @@ def main ():
     params.ip_protocol = args.ip_protocol
     params.timeout = args.timeout
     params.file = args.file
-
-    board = BoardShim (args.board_id, params)
-    sampling_rate = BoardShim.get_sampling_rate (args.board_id)
```
Comment (Andrey1994) [inline tests/python/eeg_metrics.py:43]: and the same here, lets move to another file. I really want to keep code samples from docs less than one screen of code

**Comment by Andrey1994** [comment]
Also, it will not work on android and its ok but we should add smth like 
```
#ifdef Android
return UNSUPPORTED....
```

**Comment by John42506176Linux** [comment]
Also here's a comment on how I addressed the less simple things in the PR

-I created a full version for eeg_metrics for the CPP and python file, the eeg_metrics only has added the master_board_id thing we talked about.

-I added everything needed for Valgrind so it should be fine now, the only thing I saw was a slight issue for the KNN classifier regarding the safe logger.

- It will work if data is less than 10 but considering all the other classifiers use it, I thought it'd be better to keep consistency.

-All of the svm_classifier have added the if_def android.

**Comment by John42506176Linux** [comment]
Just fixed some minor log issues valgrind runs perfectly now.
ValgrindTest

**Comment by Andrey1994** [comment]
lets revert commit https://github.com/brainflow-dev/brainflow/pull/112/commits/e9a42c2c7e2c7556d6c77ad7914caebf623081d3 and make it like here https://github.com/John42506176Linux/brainflow/pull/2 probably in all bindings

**Comment by Andrey1994** [comment]
CI is down due to an error in java test

**Comment by Andrey1994** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project brainflow: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/brainflow-dev/brainflow/java-package/brainflow/src/main/java/brainflow/examples/EEGMetrics.java:[54,30] variable model_params is already defined in method main(java.lang.String[])
[ERROR] /home/travis/build/brainflow-dev/brainflow/java-package/brainflow/src/main/java/brainflow/examples/EEGMetrics.java:[56,17] variable concentration is already defined in method main(java.lang.String[])
```

**Pull request merged**